src/backend/distributed/planner/local_distributed_join_planner.c

```diff
@@ -0,0 +1,513 @@
+#else
+#include "nodes/relation.h"
+#endif
+#include "utils/builtins.h"
+#include "utils/guc.h"
+#include "utils/lsyscache.h"
+
+#define INVALID_RTE_IDENTITY -1
+
+/*
+ * Managed via a GUC
+ */
+int LocalTableJoinPolicy = LOCAL_JOIN_POLICY_AUTO;
+
+typedef struct RangeTableEntryDetails
+{
+	RangeTblEntry *rangeTableEntry;
+	List *requiredAttributeNumbers;
+	bool hasConstantFilterOnUniqueColumn;
+} RangeTableEntryDetails;
+
+typedef struct ConversionCandidates
+{
+	List *distributedTableList; /* reference or distributed table */
+	List *localTableList; /* local or citus local table */
+}ConversionCandidates;
+
+typedef struct IndexColumns
+{
+	List *indexColumnNos;
+}IndexColumns;
+
+typedef enum ConversionChoice
+{
+	CONVERT_LOCAL_TABLES = 1,
+	CONVERT_DISTRIBUTED_TABLES = 2
+}ConversionChoice;
+
+static bool HasConstantFilterOnUniqueColumn(RangeTblEntry *rangeTableEntry,
+											RelationRestriction *relationRestriction);
+static List * RequiredAttrNumbersForRelation(RangeTblEntry *relationRte,
+											 PlannerRestrictionContext *
+											 plannerRestrictionContext);
+static ConversionCandidates * CreateConversionCandidates(PlannerRestrictionContext *
+														 plannerRestrictionContext,
+														 List *rangeTableList,
+														 int resultRTEIdentity);
+static void AppendUniqueIndexColumnsToList(Form_pg_index indexForm, List **uniqueIndexes);
+static ConversionChoice GetConversionChoice(ConversionCandidates *
+											conversionCandidates,
+											PlannerRestrictionContext *
+											plannerRestrictionContext);
+static bool AllRangeTableEntriesHaveUniqueIndex(List *rangeTableEntryDetailsList);
+static bool FirstIsSuperSetOfSecond(List *firstIntList, List *secondIntList);
+static void ConvertRTEsToSubquery(List *rangeTableEntryDetailsList,
+								  RecursivePlanningContext *context);
+
+/*
+ * RecursivelyPlanLocalTableJoins gets a query and the planner
+ * restrictions. As long as the query is not plannable by router planner,
+ * it converts either a local or distributed table to a subquery.
+ */
+void
+RecursivelyPlanLocalTableJoins(Query *query,
+							   RecursivePlanningContext *context, List *rangeTableList)
+{
+	PlannerRestrictionContext *plannerRestrictionContext =
+		GetPlannerRestrictionContext(context);
+
+	int resultRTEIdentity = INVALID_RTE_IDENTITY;
+	if (IsModifyCommand(query))
+	{
+		RangeTblEntry *resultRTE = ExtractResultRelationRTE(query);
+		resultRTEIdentity = GetRTEIdentity(resultRTE);
+	}
+	ConversionCandidates *conversionCandidates =
+		CreateConversionCandidates(plannerRestrictionContext,
+								   rangeTableList, resultRTEIdentity);
+
+	ConversionChoice conversionChoise =
+		GetConversionChoice(conversionCandidates, plannerRestrictionContext);
+
+	if (conversionChoise == CONVERT_LOCAL_TABLES)
+	{
+		ConvertRTEsToSubquery(conversionCandidates->localTableList, context);
+	}
+	else
+	{
+		ConvertRTEsToSubquery(conversionCandidates->distributedTableList, context);
```

Comment:
Maybe just set the list of to-be-converted RTEs in this if-else and call ConvertRTEsToSubquery at the end. That also allows for some more flexibility around what to put in the list.